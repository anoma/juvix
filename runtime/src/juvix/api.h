#ifndef JUVIX_API_H
#define JUVIX_API_H

#include <juvix/funcall.h>
#include <juvix/info.h>
#include <juvix/io.h>
#include <juvix/mem.h>
#include <juvix/object.h>

#define JUVIX_INIT \
    MEM_INIT;      \
    funcall_init()

// MAX_ARGS is the maximum number of function arguments plus the maximum number
// of supplied arguments, at least 3. For example, if the program contains `I I
// I I 1` where `I` is a one-argument identity function, then MAX_ARGS is
// 5.
#define JUVIX_PROLOGUE(MAX_ARGS)                    \
    STATIC_ASSERT(MAX_ARGS >= 3);                   \
    MEM_DECLS;                                      \
    UNUSED word_t juvix_temp_var;                   \
    FUNCALL_DECLS(MAX_ARGS);                        \
    JUVIX_INIT;                                     \
    STACK_PUSH_ADDR(LABEL_ADDR(juvix_program_end)); \
    goto juvix_program_start;                       \
    DECL_CALL_CLOSURES;                             \
    DECL_APPLY_CLOSURE;                             \
    DECL_APPLY_1;                                   \
    DECL_TAIL_APPLY_1;                              \
    DECL_APPLY_2;                                   \
    DECL_TAIL_APPLY_2;                              \
    DECL_APPLY_3;                                   \
    DECL_TAIL_APPLY_3;                              \
    juvix_program_start:

#define JUVIX_EPILOGUE \
    juvix_program_end: \
    STACK_POPT;        \
    IO_INTERPRET;      \
    io_print_toplevel(juvix_result);

// Temporary / local vars
#define DECL_TMP(k) UNUSED word_t juvix_tmp_##k
#define TMP(k) juvix_tmp_##k

// Begin a function definition. `max_stack` is the maximum stack allocation in
// the function.
#define JUVIX_FUNCTION(label, max_stack) \
    label:                               \
    STACK_ENTER((max_stack))

/*
    Macro sequence for function definition:

closure_label:
    ARG(0) = CARG(0);
    ...
    ARG(m) = CARG(m);
    JUVIX_FUNCTION(function_label, max_stack);
    {
        DECL_TMP(0);
        DECL_TMP(1);
        ...
        DECL_TMP(max_tmp-1);
        <code>
    }
*/

// Begin a function with no stack allocation.
#define JUVIX_FUNCTION_NS(label) \
    label:

#define JUVIX_INT_ADD(var0, var1, var2) (var0 = smallint_add(var1, var2))
#define JUVIX_INT_SUB(var0, var1, var2) (var0 = smallint_sub(var1, var2))
#define JUVIX_INT_MUL(var0, var1, var2) (var0 = smallint_mul(var1, var2))
#define JUVIX_INT_DIV(var0, var1, var2) (var0 = smallint_div(var1, var2))
#define JUVIX_INT_MOD(var0, var1, var2) (var0 = smallint_mod(var1, var2))

#define JUVIX_INT_LT(var0, var1, var2) (var0 = smallint_lt(var1, var2))
#define JUVIX_INT_LE(var0, var1, var2) (var0 = smallint_le(var1, var2))
#define JUVIX_INT_EQ(var0, var1, var2) (var0 = make_bool(var1 == var2))

#define JUVIX_VAL_EQ(var0, var1, var2) \
    (var0 = make_bool(juvix_equal(var1, var2)))

#define JUVIX_STR_CONCAT(var0, var1, var2) CONCAT_CSTRINGS(var0, var1, var2)

#define JUVIX_STR_TO_INT(var0, var1) \
    (var0 = make_smallint(strtoint(get_cstring(var1))))

#define JUVIX_SHOW(var0, var1)                                     \
    do {                                                           \
        char *juvix_buf = print(var1);                             \
        if (strlen(juvix_buf) / sizeof(word_t) + 1 > MAX_FIELDS) { \
            error_exit_msg("show: string too long");               \
        }                                                          \
        ALLOC_CSTRING(var0, juvix_buf);                            \
        free_strbuf(juvix_buf);                                    \
    } while (0)

#define JUVIX_ASSIGN(var0, val) (var0 = val)

#define JUVIX_TRACE(val) (io_trace(val))
#define JUVIX_DUMP (stacktrace_dump())
#define JUVIX_FAILURE(val) \
    do {                   \
        io_trace(val);     \
        error_exit();      \
    } while (0)

#define JUVIX_ARGS_NUM(var, val) (var = make_smallint(get_closure_largs(val)))

#define JUVIX_ALLOC_INT(var, val) (var = make_smallint(val))
// ALLOC_CONSTR_BOXED(var, uid, nargs)
// ALLOC_CONSTR_BOXED_TAG(var, uid)
// ALLOC_CONSTR_UNBOXED(var, uid)
// ALLOC_CONSTR_PAIR(var)

// ALLOC_CSTRING(var, str)

// ALLOC_CLOSURE(var, fuid, addr, nargs, largs)
// EXTEND_CLOSURE(dest, src, n)

// CALL(fuid, fun_label, return_label)
// TAIL_CALL(fuid, fun_label)
// CALL_CLOSURE(cl)
// TAIL_CALL_CLOSURE(cl)
// CALL_CLOSURES(nargs, return_label)
// TAIL_CALL_CLOSURES(nargs)
// RETURN

#define JUVIX_BRANCH(val, CODE_TRUE, CODE_FALSE) \
    if (is_true(val)) {                          \
        CODE_TRUE;                               \
    } else {                                     \
        CODE_FALSE;                              \
    }

/*
    Compilation of the `Case` instruction:

    switch (var) {
        case UNBOXED_CONSTR_HEADER_1: {<code>}; break;
        ...
        case UNBOXED_CONSTR_HEADER_N: {<code>}; break;
        default:
            switch (get_header(var)) {
                case BOXED_CONSTR_HEADER_1: {<code>}; break;
                ...
                case BOXED_CONSTR_HEADER_M: {<code>}; break;
                default:
                    <code> (or UNREACHABLE)
            }
    }

    Single-branch switches shouldn't be output.
*/

#endif
