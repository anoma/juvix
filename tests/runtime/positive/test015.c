/* Branching, matching and recursion on lists */

#include <juvix/api.h>

#define JUVIX_DECL_ARGS \
    DECL_REG_ARG(0);    \
    DECL_REG_ARG(1)

#define CONSTRS_NUM (BUILTIN_UIDS_NUM + 2)

static constr_info_t juvix_constr_info_array[CONSTRS_NUM] = {
    BUILTIN_UIDS_INFO, {"nil"}, {"cons"}};

#define UID_NIL FIRST_USER_UID
#define UID_CONS (FIRST_USER_UID + 1)

#define CONSTR_NIL MAKE_HEADER(UID_NIL, 0)

int main() {
    JUVIX_PROLOGUE(2, JUVIX_DECL_ARGS);

    juvix_constrs_num = CONSTRS_NUM;
    juvix_constr_info = juvix_constr_info_array;

    CALL(0, juvix_function_main, juvix_label_0);
    goto juvix_program_end;

    JUVIX_FUNCTION_NS(juvix_function_hd);
    {
        juvix_result = CONSTR_ARG(ARG(0), 0);
        RETURN_NS;
    }

    JUVIX_FUNCTION_NS(juvix_function_tl);
    {
        juvix_result = CONSTR_ARG(ARG(0), 1);
        RETURN_NS;
    }

    JUVIX_FUNCTION_NS(juvix_function_null);
    {
        switch (ARG(0)) {
            case CONSTR_NIL:
                juvix_result = BOOL_TRUE;
                break;
            default:  // cons
                juvix_result = BOOL_FALSE;
                break;
        }
        RETURN_NS;
    }

    JUVIX_FUNCTION(juvix_function_map, 3);
    {
        switch (ARG(1)) {
            case CONSTR_NIL:
                juvix_result = CONSTR_NIL;
                RETURN;
            default: {
                DECL_TMP(0);
                DECL_TMP(1);
                TMP(0) = ARG(0);
                TMP(1) = ARG(1);
                STACK_PUSH(TMP(0));
                STACK_PUSH(TMP(1));
                ARG(0) = TMP(0);
                ARG(1) = CONSTR_ARG(TMP(1), 1);
                CALL(0, juvix_function_map, juvix_label_1);
                STACK_POP(TMP(1));
                STACK_POP(TMP(0));
                STACK_PUSH(juvix_result);
                ASSIGN_CARGS(TMP(0), {
                    CARG(juvix_closure_nargs) = CONSTR_ARG(TMP(1), 0);
                });
                CALL_CLOSURE(TMP(0), juvix_label_2);
                STACK_POP(TMP(0));
                TMP(1) = juvix_result;
                PREALLOC(
                    3,
                    {
                        STACK_PUSH(TMP(0));
                        STACK_PUSH(TMP(1));
                    },
                    {
                        STACK_POP(TMP(1));
                        STACK_POP(TMP(0));
                    });
                ALLOC_CONSTR_BOXED(juvix_result, UID_CONS, 2);
                CONSTR_ARG(juvix_result, 0) = TMP(1);
                CONSTR_ARG(juvix_result, 1) = TMP(0);
                RETURN;
            }
        }
    }

    JUVIX_FUNCTION(juvix_function_map_1, 3);
    {
        STACK_PUSH(ARG(0));
        STACK_PUSH(ARG(1));
        ARG(0) = ARG(1);
        CALL(0, juvix_function_null, juvix_label_3);
        STACK_POP(ARG(1));
        STACK_POP(ARG(0));
        JUVIX_BRANCH(
            juvix_result, { juvix_result = CONSTR_NIL; },
            {
                DECL_TMP(0);
                DECL_TMP(1);
                STACK_PUSH(ARG(0));
                ARG(0) = ARG(1);
                CALL(0, juvix_function_tl, juvix_label_4);
                ARG(0) = STACK_TOP;
                STACK_PUSH(ARG(1));
                ARG(1) = juvix_result;
                CALL(0, juvix_function_map_1, juvix_label_5);
                STACK_POP(TMP(1));
                STACK_POP(TMP(0));
                STACK_PUSH(TMP(0));
                STACK_PUSH(juvix_result);
                ARG(0) = TMP(1);
                CALL(0, juvix_function_hd, juvix_label_6);
                STACK_POP(TMP(1));  // TMP(1) = map tail
                STACK_POP(TMP(0));  // TMP(0) = function
                STACK_PUSH(TMP(1));
                ASSIGN_CARGS(TMP(0),
                             { CARG(juvix_closure_nargs) = juvix_result; });
                CALL_CLOSURE(TMP(0), juvix_label_7);
                STACK_POP(TMP(1));
                PREALLOC(
                    3, { STACK_PUSH(TMP(1)); }, { STACK_POP(TMP(1)); });
                ALLOC_CONSTR_BOXED(TMP(0), UID_CONS, 2);
                CONSTR_ARG(TMP(0), 0) = juvix_result;
                CONSTR_ARG(TMP(0), 1) = TMP(1);
                juvix_result = TMP(0);
            });
        RETURN;
    }

juvix_closure_add_one:
    ARG(0) = CARG(0);
    JUVIX_FUNCTION_NS(juvix_function_add_one);
    {
        JUVIX_INT_ADD(juvix_result, ARG(0), make_smallint(1));
        RETURN_NS;
    }

    JUVIX_FUNCTION(juvix_function_gen, 2);
    {
        DECL_TMP(0);
        JUVIX_VAL_EQ(TMP(0), ARG(0), make_smallint(0));
        JUVIX_BRANCH(
            TMP(0),
            {
                juvix_result = CONSTR_NIL;
                RETURN;
            },
            {
                JUVIX_INT_SUB(TMP(0), ARG(0), make_smallint(1));
                STACK_PUSH(ARG(0));
                ARG(0) = TMP(0);
                CALL(0, juvix_function_gen, juvix_label_gen_1);
                STACK_POP(ARG(0));
                PREALLOC(
                    3, { STACK_PUSH(ARG(0)); }, { STACK_POP(ARG(0)); });
                ALLOC_CONSTR_BOXED(TMP(0), UID_CONS, 2);
                CONSTR_ARG(TMP(0), 0) = ARG(0);
                CONSTR_ARG(TMP(0), 1) = juvix_result;
                juvix_result = TMP(0);
                RETURN;
            });
    }

    JUVIX_FUNCTION(juvix_function_sum, 2);
    {
        switch (ARG(0)) {
            case CONSTR_NIL:
                juvix_result = make_smallint(0);
                RETURN;
            default: {
                DECL_TMP(0);
                DECL_TMP(1);
                TMP(0) = CONSTR_ARG(ARG(0), 0);
                STACK_PUSH(TMP(0));
                ARG(0) = CONSTR_ARG(ARG(0), 1);
                CALL(0, juvix_function_sum, juvix_label_sum_1);
                STACK_POP(TMP(0));
                JUVIX_INT_ADD(TMP(1), TMP(0), juvix_result);
                juvix_result = TMP(1);
                RETURN;
            }
        }
    }

    JUVIX_FUNCTION(juvix_function_length, 1);
    {
        switch (ARG(0)) {
            case CONSTR_NIL:
                juvix_result = make_smallint(0);
                RETURN;
            default: {
                DECL_TMP(0);
                ARG(0) = CONSTR_ARG(ARG(0), 1);
                CALL(0, juvix_function_length, juvix_label_length_1);
                JUVIX_INT_ADD(TMP(0), juvix_result, make_smallint(1));
                juvix_result = TMP(0);
                RETURN;
            }
        }
    }

    JUVIX_FUNCTION_NS(juvix_function_length_tail);
    {
        switch (ARG(0)) {
            case CONSTR_NIL:
                juvix_result = ARG(1);
                RETURN_NS;
            default: {
                ARG(0) = CONSTR_ARG(ARG(0), 1);
                JUVIX_INT_ADD(ARG(1), ARG(1), make_smallint(1));
                TAIL_CALL_NS(0, juvix_function_length_tail);
            }
        }
    }

    JUVIX_FUNCTION(juvix_function_main, MAX_STACK_DELTA);
    {
        DECL_TMP(0);  // cons 0 (cons 1 nil)
        DECL_TMP(1);  // cons 1 nil
        DECL_TMP(2);  // calloc add_one 0
        ALLOC_CONSTR_BOXED(TMP(1), UID_CONS, 2);
        CONSTR_ARG(TMP(1), 0) = make_smallint(1);
        CONSTR_ARG(TMP(1), 1) = CONSTR_NIL;
        ALLOC_CONSTR_BOXED(TMP(0), UID_CONS, 2);
        CONSTR_ARG(TMP(0), 0) = make_smallint(0);
        CONSTR_ARG(TMP(0), 1) = TMP(1);
        ALLOC_CLOSURE(TMP(2), 0, LABEL_ADDR(juvix_closure_add_one), 0, 1);
        ARG(0) = TMP(2);
        ARG(1) = TMP(0);
        CALL(0, juvix_function_map, juvix_label_8);
        JUVIX_TRACE(juvix_result);
        ARG(0) = TMP(2);
        ARG(1) = TMP(0);
        CALL(0, juvix_function_map_1, juvix_label_9);
        JUVIX_TRACE(juvix_result);
        ARG(0) = make_smallint(10000);
        CALL(0, juvix_function_gen, juvix_label_10);
        ARG(0) = TMP(2);
        ARG(1) = juvix_result;
        CALL(0, juvix_function_map, juvix_label_11);
        ARG(0) = juvix_result;
        CALL(0, juvix_function_sum, juvix_label_12);
        JUVIX_TRACE(juvix_result);
        ARG(0) = make_smallint(1000000);
        CALL(0, juvix_function_gen, juvix_label_13);
        STACK_PUSH(juvix_result);
        ARG(0) = TMP(2);
        ARG(1) = juvix_result;
        CALL(0, juvix_function_map, juvix_label_14);
        ARG(0) = juvix_result;
        CALL(0, juvix_function_length, juvix_label_15);
        JUVIX_TRACE(juvix_result);
        STACK_POP(ARG(0));
        ARG(1) = make_smallint(0);
        TAIL_CALL(0, juvix_function_length_tail);
    }

    JUVIX_EPILOGUE;
    return 0;
}
