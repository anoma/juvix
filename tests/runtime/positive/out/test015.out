cons 1 (cons 2 nil)
cons 1 (cons 2 nil)
50015000
1000000
1000000
